### PR TITLE
Calendar: Multiple Category Assignment

### PR DESCRIPTION
**Pull request by retokohli**
Note the database migration script with instructions:
modules/Calendar/Model/Data/migrate-calendar-event-to-category-relation.sql

### CODE REVIEW

File: modules/Calendar/Controller/CalendarEvent.class.php [hunk 1]
```diff
@@ -2530,27 +2510,22 @@ public function getEventEntity($id, $formDatas= array())
...
         foreach ($formDatas['fields'] as $columnName => $columnValue) {
             $fieldName  = $classMetaData->getFieldName($columnName);
-            if ($fieldName == 'catid') {
-                $fieldName   = 'category';
-                $columnValue = $this
-                    ->em
-                    ->getRepository('Cx\Modules\Calendar\Model\Entity\Category')
-                    ->findOneById($columnValue);
+// TODO: Test! Can Events be synchronized by using the categories collection?
+            if ($fieldName === 'categories') {
```
Comment (daeppen) [inline modules/Calendar/Controller/CalendarEvent.class.php]: this won't work as 'categories' is no longer a regular field

File: modules/Calendar/Model/Entity/Event.class.php [hunk 2]
```diff
@@ -2006,23 +2007,25 @@ public function getRegistrations()
...
+     * @param \Doctrine\Common\Collections\Collection $categories
+     * @return Event
+     * @author Reto Kohli <reto.kohli@comvation.com>
      */
-    public function setCategory(\Cx\Modules\Calendar\Model\Entity\Category $category)
+    public function setCategories(\Doctrine\Common\Collections\Collection $categories)
     {
-        $this->category = $category;
+        $this->categories[] = $categories;
+        return $this;
```
Comment (daeppen) [inline modules/Calendar/Model/Entity/Event.class.php]: `return $this;`
returing the object of a model setter does not comply with the model standard. Please remove this return operation

File: modules/Calendar/Model/Entity/Event.class.php [hunk 3]
```diff
@@ -2006,23 +2007,25 @@ public function getRegistrations()
...
+     * Set categories
+     * @param \Doctrine\Common\Collections\Collection $categories
+     * @return Event
+     * @author Reto Kohli <reto.kohli@comvation.com>
      */
-    public function setCategory(\Cx\Modules\Calendar\Model\Entity\Category $category)
+    public function setCategories(\Doctrine\Common\Collections\Collection $categories)
     {
-        $this->category = $category;
+        $this->categories[] = $categories;
```
Comment (daeppen) [inline modules/Calendar/Model/Entity/Event.class.php]: this should be as follows, right (without brackets)?
`$this->categories = $categories;`

File: modules/Calendar/Model/Data/migrate-calendar-event-to-category-relation.sql [hunk 4]
```diff
@@ -0,0 +1,18 @@
+-- Run the migration queries first:
```
Comment (daeppen) [inline modules/Calendar/Model/Data/migrate-calendar-event-to-category-relation.sql]: Please add the migration statements to the description of the PR. Then, please drop this file.
Note: please do also remove any charset/collation specifications in the migration statements.

File: modules/Calendar/View/Template/Backend/module_calendar_modify_event.html [hunk 5]
```diff
@@ -776,9 +776,8 @@
         cx.jQuery("#calendar_event_invite_crm_memberships").chosen({ width: '250px' });
     });
 </script>
-
-
-<form action="index.php?cmd={CALENDAR_MODULE_NAME}&act=modify_event" id="formModifyEvent" name="formModifyEvent" method="post" onsubmit="selectAll(document.formModifyEvent.elements['selectedGroups[]']); {CALENDAR_EVENT_ONSUBMIT_PUBLICATIONS} return formCheck();">
+<form action="index.php?cmd={CALENDAR_MODULE_NAME}&act=modify_event" id="formModifyEvent" name="formModifyEvent" method="post" onsubmit="selectAll(document.formModifyEvent.elements['selectedGroups[]']); {
+      CALENDAR_EVENT_ONSUBMIT_PUBLICATIONS} return formCheck();">
```
Comment (daeppen) [inline modules/Calendar/View/Template/Backend/module_calendar_modify_event.html]: this white-space change seems to break the sigma placeholder

File: modules/Calendar/Model/Data/migrate-calendar-event-to-category-relation.sql (hunk 4 above)
Comment (retokohli) [inline modules/Calendar/Model/Data/migrate-calendar-event-to-category-relation.sql]: Fixed: 6e3ed6f

File: modules/Calendar/Model/Entity/Event.class.php (hunk 3 above)
Comment (retokohli) [inline modules/Calendar/Model/Entity/Event.class.php]: Fixed: 1177406

File: modules/Calendar/Model/Entity/Event.class.php (hunk 2 above)
Comment (retokohli) [inline modules/Calendar/Model/Entity/Event.class.php]: Fixed: bb3f647

File: modules/Calendar/View/Template/Backend/module_calendar_modify_event.html (hunk 5 above)
Comment (retokohli) [inline modules/Calendar/View/Template/Backend/module_calendar_modify_event.html]: Fixed: 343c58d

File: modules/Calendar/Controller/CalendarEvent.class.php (hunk 1 above)
Comment (retokohli) [inline modules/Calendar/Controller/CalendarEvent.class.php]: I don't know how to properly test (and thus fix) this.
How would you have to treat this then, as a "non-regular" field?

**Pull request closed without merging**